client/allocrunner/consul_hook.go

```diff
@@ -97,6 +97,7 @@ func (h *consulHook) Prerun() error {
 
 	err := mErr.ErrorOrNil()
 	if err != nil {
+		multierror.Flatten(err)
```

Comment:
I don't remember using this method before, but it doesn't seem to be doing any for us here?

Here's an error with 1.7.0-beta.2:

![image](https://github.com/hashicorp/nomad/assets/775380/09a0b50d-3f7f-4113-8f50-0258ff6dd22b)

```
    2023-11-21T20:30:54.242-0500 [ERROR] client.alloc_runner: prerun failed: alloc_id=7a5af945-ce5a-d6f6-859c-ef013cf44420
  error=
  | pre-run hook "consul" failed: 1 error occurred:
  | \t* 1 error occurred:
  | \t* 1 error occurred:
  | \t* failed to authenticate with consul for identity consul-service_mongo-db: Unexpected response code: 403 (ACL not found: auth method "nomad-services" not found)
  |
  |
  |
  |
  |
```

And the error shows up the same way in this branch:

![image](https://github.com/hashicorp/nomad/assets/775380/bd90b8a1-88af-4783-8926-3b12aeef9337)

```
    2023-11-21T20:32:59.484-0500 [ERROR] client.alloc_runner: prerun failed: alloc_id=7d4984f7-3deb-3dc3-5850-9669eaf99caf
  error=
  | pre-run hook "consul" failed: 1 error occurred:
  | \t* 1 error occurred:
  | \t* 1 error occurred:
  | \t* failed to authenticate with consul for identity consul-service_mongo-db: Unexpected response code: 403 (ACL not found: auth method "nomad-services" not found)
  |
  |
  |
  |
  |
```

I haven't looked too much into the code yet, but it seems like the error multierror topology is something like this:

```mermaid
graph TB
  RootError((Root Error)) --> GroupServicesError
  RootError --> Task

  GroupServicesError((Group Services Errors)) --> E1((Error 1))
  GroupServicesError --> E2((Error 2))
  GroupServicesError --> E3((Error 3))

  Task((Task Errors)) --> TaskServicesError
  Task --> TaskErrors

  TaskServicesError((Tasks Services Errors)) --> E4((Error 1))
  TaskServicesError --> E5((Error 2))
  TaskServicesError --> E6((Error 3))

  TaskErrors((Tasks Errors)) --> E7((Error 1))
  TaskErrors --> E8((Error 2))
  TaskErrors --> E9((Error 3))
```

So instead of flattening we probably want to accumulate errors at the root with the proper context. For example, error above would be just:

```
    2023-11-21T20:32:59.484-0500 [ERROR] client.alloc_runner: prerun failed: alloc_id=7d4984f7-3deb-3dc3-5850-9669eaf99caf
  error=
  | pre-run hook "consul" failed: 1 error occurred:
  | \t* failed to authenticate with consul for identity consul-service_mongo-db: Unexpected response code: 403 (ACL not found: auth method "nomad-services" not found)
```

I think `errors.Unwrap()` may skip a level, but I haven't fully gotten used to how this works 😅 

Otherwise you could do type assertion at the caller.

Alternatively, since the `prepare*` methods are internal, they could return a `multierror.Error` directly and the caller merges all internal errors into one.

Reply:
oof, apologies for not checking this thoroughly before :/ 

4767ee0 changes the implementation using `multierror.Error` returns for internal methods and type assertion for the returned error of consul client. Results in a much neater error message:

```
    2023-11-23T17:37:46.487+0100 [ERROR] client.alloc_runner: prerun failed: alloc_id=1e22a6a9-4b49-6fdd-f2c8-0d65b5685d55
  error=
  | pre-run hook "consul" failed: 1 error occurred:
  | \t* failed to authenticate with consul for identity consul_default: Unexpected response code: 403 (ACL not found: auth method "nomad-tasks" not found)
  |
```